subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/PmdPluginBuildCacheIntegrationTest.groovy

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.quality
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
+
+import static org.hamcrest.Matchers.containsString
+import static org.hamcrest.Matchers.not
+
+class PmdPluginBuildCacheIntegrationTest extends AbstractIntegrationSpec implements DirectoryBuildCacheFixture {
+
+    def setup() {
+        buildFile << """
+            apply plugin: "java"
+            apply plugin: "pmd"
+            pmd {
+                incrementalAnalysis = true
+            }
+
+            ${mavenCentralRepository()}
+
+            ${!TestPrecondition.FIX_TO_WORK_ON_JAVA9.fulfilled ? "sourceCompatibility = 1.6" : ""}
+        """.stripIndent()
+    }
+
+    @Requires(TestPrecondition.NOT_WINDOWS)
+    def "pmd can work with build cache"() {
+        given:
+        // put into cache
+        goodCode()
+        succeeds('pmdMain')
+
+        when:
+        // cache miss
+        badCode()
+        fails('pmdMain')
+
+        then:
+        file("build/reports/pmd/main.xml").assertContents(containsString('BadClass'))
+
+        when:
+        // cache hit
+        removeBadCode()
```

Comment:
❌ Since the analysis cache is on a class basis, it would probably be better to change the offending class and fix the problem instead of removing it. Please have another look at https://github.com/gradle/gradle/blob/a16a105cd2b27a8c49d6f1621e18fb0cd3410343/subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/CachedScalaCompileIntegrationTest.groovy#L137-L179.

Reply:
The bad code is removed to hit the cache, since I think the build cache key will contain the source code content hash value, right?

Reply:
Yes, you are removing the bad code file to get back to the previous input state to have a build cache hit. What I am saying is that instead of introducing a new file with pmd violations it would probably be better to introduce the violations in an already existing file and then revert the change again to obtain the cache hit.

Reply:
Oh, brilliant idea, thanks!